contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -3937,3 +3937,155 @@ update_db_owner(const char *new_owner_name, const char *db_name)
+	ScanKeyData scanKey[2];
+	int16		dbid = get_db_id(dbname);
+	int16		old_dbid;
+	const char *old_dbname;
+	const char	*prev_current_user;
+	MigrationMode baseline_mode = is_user_database_singledb(dbname) ? SINGLE_DB : MULTI_DB;
+
+	/* Fetch the relation */
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+									AccessShareLock);
+	dsc = RelationGetDescr(bbf_schema_rel);
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_object_type,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_TEXTEQ,
+				CStringGetTextDatum(OBJ_SCHEMA));
+
+	scan = systable_beginscan(bbf_schema_rel, get_bbf_schema_perms_idx_oid(),
+							true, NULL, 2, scanKey);
+	tuple_bbf_schema = systable_getnext(scan);
+
+	/* Set current user to session user for create permissions */
+	prev_current_user = GetUserNameFromId(GetUserId(), false);
+	bbf_set_current_user("sysadmin");
+
+	old_dbid = get_cur_db_id();
+	old_dbname = get_cur_db_name();
+	set_cur_db(dbid, dbname);	/* temporarily set current dbid as the new id */
+
+	while (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		bool		isnull;
+		const char	*schema_name;
+		const char	*grantee;
+		int			current_permission;
+		const char	*schema_owner;
+		const char	*physical_schema;
+		const char	*dbo_user;
+		const char	*db_owner;
+		int			i;
+
+		schema_name = TextDatumGetCString(heap_getattr(tuple_bbf_schema, Anum_bbf_schema_perms_schema_name, dsc, &isnull));
+		grantee = TextDatumGetCString(heap_getattr(tuple_bbf_schema, Anum_bbf_schema_perms_grantee, dsc, &isnull));
+		current_permission = DatumGetInt32(heap_getattr(tuple_bbf_schema, Anum_bbf_schema_perms_permission, dsc, &isnull));
+
+		physical_schema = get_physical_schema_name_by_mode((char *) dbname, schema_name, baseline_mode);
+		dbo_user = get_dbo_role_name_by_mode((char *)dbname, baseline_mode);
+		db_owner = get_db_owner_name_by_mode((char *) dbname, baseline_mode);
+		schema_owner = GetUserNameFromId(get_owner_of_schema(physical_schema), false);
+
+		/* If schema owner is other that dbo or db_owner user, only then execute ALTER DEFAULT PRIVILEGES. */
+		if ((strcmp(schema_owner, dbo_user) != 0) && (strcmp(schema_owner, db_owner) != 0))
+		{
+			/* For each permission, grant alter default privileges explicitly. */
+			for (i = 0; i < NUMBER_OF_PERMISSIONS; i++)
+			{
+				if ((current_permission & permissions[i]) &&  permissions[i] != ACL_EXECUTE)
+				{
+					PG_TRY();
+					{
+						const char	*query = NULL;
+						List		*res;
+						AlterDefaultPrivilegesStmt	*alterdefpriv;
+						PlannedStmt	*wrapper;
+						query = psprintf("ALTER DEFAULT PRIVILEGES FOR ROLE %s, %s IN SCHEMA %s GRANT %s ON TABLES TO %s", dbo_user, schema_owner, physical_schema, privilege_to_string(permissions[i]), grantee);
+						res = raw_parser(query, RAW_PARSE_DEFAULT);
```

Comment:
We need to execute `GRANT ... ON ALL TABLES` well to handle the older objects created by u1 user(referring the example given in commit desc.) after granting the schema privs. in older version.
Pseudo Code of example:
```
-- Old version
create schema sch1 owner u1
grant privs on schema to u2
create table t1()
t1 will not be accessible to u2 even if it got schema perms.
-- Latest Version
t1 should be accessible to u2
```

Reply:
Please add testcase for the same.